### PR TITLE
Planner Bug: Joins inside derived table

### LINKED ISSUE #14993: Joins inside derived tables
A query like this:

```sql
SELECT count(*)
FROM (SELECT DISTINCT u.user_id
      FROM user u
               JOIN user_extra ue ON u.id = ue.user_id
               JOIN music m ON m.id = u.id) dt
```

Will be planned like this by the vtgate planner.

```json
{
  "QueryType": "SELECT",
  "Original": "SELECT count(*) FROM (SELECT DISTINCT u.user_id FROM user u JOIN user_extra ue ON u.id = ue.user_id JOIN music m ON m.id = u.id) dt",
  "Instructions": {
    "OperatorType": "Aggregate",
    "Variant": "Scalar",
    "Aggregates": "count_star(0) AS count(*)",
    "Inputs": [
      {
        "OperatorType": "SimpleProjection",
        "Columns": [
          1
        ],
        "Inputs": [
          {
            "OperatorType": "Distinct",
            "Collations": [
              "(0:2)",
              "1"
            ],
            "Inputs": [
              {
                "OperatorType": "Join",
                "Variant": "Join",
                "JoinColumnIndexes": "R:0,L:1,R:1",
                "JoinVars": {
                  "m_id": 2
                },
                "TableName": "music_`user`, user_extra",
                "Inputs": [
                  {
                    "OperatorType": "Route",
                    "Variant": "Scatter",
                    "Keyspace": {
                      "Name": "user",
                      "Sharded": true
                    },
                    "FieldQuery": "select m.id, 1, dt.`m.id` from (select m.id as `m.id` from music as m where 1 != 1) as dt where 1 != 1",
                    "Query": "select distinct m.id, 1, dt.`m.id` from (select m.id as `m.id` from music as m) as dt",
                    "Table": "music"
                  },
                  {
                    "OperatorType": "Route",
                    "Variant": "EqualUnique",
                    "Keyspace": {
                      "Name": "user",
                      "Sharded": true
                    },
                    "FieldQuery": "select u.user_id, weight_string(u.user_id) from (select u.user_id from `user` as u, user_extra as ue where 1 != 1) as dt where 1 != 1",
                    "Query": "select distinct u.user_id, weight_string(u.user_id) from (select u.user_id from `user` as u, user_extra as ue where u.id = :m_id and u.id = ue.user_id) as dt",
                    "Table": "`user`, user_extra",
                    "Values": [
                      ":m_id"
                    ],
                    "Vindex": "user_index"
                  }
                ]
              }
            ]
          }
        ]
      }
    ]
  },
  "TablesUsed": [
    "user.music",
    "user.user",
    "user.user_extra"
  ]
}
```

Both queries we are sending down to the tablets are obviously wrong:

```sql
-- left
select distinct m.id, 1, dt.`m.id` from (select m.id as `m.id` from music as m) as dt;

--right
select distinct u.user_id, weight_string(u.user_id)
from (select u.user_id
      from `user` as u,
           user_extra as ue
      where u.id = :m_id
        and u.id = ue.user_id) as dt
```

We are using the internal tables names on the outside of the derived table, and this will just make MySQL throw an error.

I've only been able to trigger the issue when there are three tables being joined, and two of these tables queries can be merged into a single one.

### PR DESCRIPTION
**Pull request by systay**
## Description
In some situations, the vtgate planner was producing really bad plans around derived tables and columns.

In this PR I reworked a lot of the logic around derived tables and their columns.

## Related Issue(s)
Fixes #14993

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/planbuilder/testdata/select_cases.json [hunk 1]
```diff
@@ -2028,72 +2028,7 @@
...
-            "Table": "user_extra"
-          }
-        ]
-      },
-      "TablesUsed": [
-        "user.user",
-        "user.user_extra"
-      ]
-    }
+    "plan": "expr cannot be translated, not supported: (select :0 from `user` limit 1)"
```
Comment (systay) [inline go/vt/vtgate/planbuilder/testdata/select_cases.json:2031]: this is not correct

File: go/vt/vtgate/planbuilder/operators/SQL_builder.go
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/SQL_builder.go:1]: @mattlord, when looking at this file in the `Files changed` tab it is a bit hard to read the code as it is full of Codecov warnings. Is there a possible way to disable those warnings or summarize them somewhere else?

Comment (frouioui) [reply go/vt/vtgate/planbuilder/operators/SQL_builder.go:1]: Well I guess they are already summarized on the Codecov website itself.

File: go/vt/vtgate/planbuilder/operators/projection_pushing.go [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+// claimUnusedAlias generates a unique alias based on the provided expression, ensuring no duplication in the projector
+func (p *projector) claimUnusedAlias(ae *sqlparser.AliasedExpr) string {
+	bare := ae.ColumnName()
+	alias := bare
+	var i int64 = 0
+
+	for slices.Index(p.columnAliases, alias) > -1 {
+		alias = bare + strconv.FormatInt(i, 10)
+		i++
+	}
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/projection_pushing.go:84]: being picky about coding style here, but you could also inline the definition and incrementation of `i` in the for loop as follows:
```suggestion
	for i := int64(0); slices.Index(p.columnAliases, alias) > -1; i++ {
		alias = bare + strconv.FormatInt(i, 10)
	}
```

File: go/vt/vtgate/planbuilder/operators/projection_pushing.go [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+	p *projector,
+	dt *DerivedTable,
+) Operator {
+	if len(p.columns) == 0 {
+		return src
+	}
+	proj := createProjection(ctx, src, "")
+	proj.Columns = AliasedProjections(p.columns)
+	if dt != nil {
+		kopy := *dt
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/projection_pushing.go:404]: we do not have a proper clone function for `DerivedTable`?

**Review by frouioui — APPROVED** [review]
Besides that failing plan test and the comment I left, it looks good to me! That's a really good enhancement / bug fix, thank you 🙏🏻 

Codecov is spanning a bunch of warnings in a lot of places in this PR, is this because of the cross-package coverage missing or is it due to actual missing plan tests?

File: go/vt/vtgate/planbuilder/testdata/from_cases.json [hunk 4]
```diff
@@ -4279,5 +4268,79 @@
...
+                      },
+                      {
+                        "OperatorType": "Route",
+                        "Variant": "EqualUnique",
+                        "Keyspace": {
+                          "Name": "user",
+                          "Sharded": true
+                        },
+                        "FieldQuery": "select subquery_for_count.user_id, subquery_for_count.`u.id`, weight_string(subquery_for_count.user_id) from (select u.user_id, u.id as `u.id` from `user` as u, user_extra as ue where 1 != 1) as subquery_for_count where 1 != 1",
+                        "Query": "select distinct subquery_for_count.user_id, subquery_for_count.`u.id`, weight_string(subquery_for_count.user_id) from (select u.user_id, u.id as `u.id` from `user` as u, user_extra as ue where u.id = :m_id and u.id = ue.user_id) as subquery_for_count",
```
Comment (systay) [inline go/vt/vtgate/planbuilder/testdata/from_cases.json:4324]: this doesn't look right to me. there is no need to return `u.id` here - it's just being used in the join predicate, nowhere else.

**Review by frouioui — APPROVED** [review]

File: go/vt/vtgate/planbuilder/operators/projection_pushing.go (hunk 3 above)
Comment (systay) [reply go/vt/vtgate/planbuilder/operators/projection_pushing.go:410]: nope. I'll add one

Comment (systay) [reply go/vt/vtgate/planbuilder/operators/projection_pushing.go:441]: hmm... I changed my mind. Don't think we need one. This is the only place we try to clone it, and it's not a pure cloning anyway

File: go/test/endtoend/vtgate/queries/subquery/subquery_test.go [hunk 5]
```diff
@@ -177,3 +175,14 @@ func TestSubqueryInAggregation(t *testing.T) {
...
+
+// TestSubqueryInDerivedTable
+func TestSubqueryInDerivedTable(t *testing.T) {
+	utils.SkipIfBinaryIsBelowVersion(t, 20, "vtgate")
+	mcmp, closer := start(t)
+	defer closer()
+
+	mcmp.Exec("INSERT INTO t1 (id1, id2) VALUES (1, 100), (2, 200), (3, 300), (4, 400), (5, 500);")
+	mcmp.Exec("INSERT INTO t2 (id3, id4) VALUES (10, 1), (20, 2), (30, 3), (40, 4), (50, 99)")
+	mcmp.Exec(`select t.a from (select t1.id2, t2.id3, (select id2 from t1 order by id2 limit 1) as a from t1 join t2 on t1.id1 = t2.id4) t`)
```
Comment (systay) [inline go/test/endtoend/vtgate/queries/subquery/subquery_test.go:188]: This last test will fail without the changes in this PR:

```
        	Error:      	Received unexpected error:
        	            	target: ks_union.-80.primary: vttablet: rpc error: code = NotFound desc = Unknown column 't2.id4' in 'field list' (errno 1054) (sqlstate 42S22) (CallerID: userData1): Sql: "select distinct t2.id4, 1, dt.`t2.id4` from (select t2.id4 as `t2.id4` from t2) as dt", BindVars: {}
        	            	target: ks_union.80-.primary: vttablet: rpc error: code = NotFound desc = Unknown column 't2.id4' in 'field list' (errno 1054) (sqlstate 42S22) (CallerID: userData1): Sql: "select distinct t2.id4, 1, dt.`t2.id4` from (select t2.id4 as `t2.id4` from t2) as dt", BindVars: {} (errno 1054) (sqlstate 42S22) during query: SELECT COUNT(*) FROM (SELECT DISTINCT t1.id1 FROM t1 JOIN t2 ON t1.id1 = t2.id4) dt
```

File: go/test/endtoend/vtgate/queries/subquery/subquery_test.go [hunk 6]
```diff
@@ -177,3 +175,16 @@ func TestSubqueryInAggregation(t *testing.T) {
...
+// are handled correctly when there are joins inside the derived table
+func TestSubqueryInDerivedTable(t *testing.T) {
+	utils.SkipIfBinaryIsBelowVersion(t, 20, "vtgate")
+	mcmp, closer := start(t)
+	defer closer()
+
+	mcmp.Exec("INSERT INTO t1 (id1, id2) VALUES (1, 100), (2, 200), (3, 300), (4, 400), (5, 500);")
+	mcmp.Exec("INSERT INTO t2 (id3, id4) VALUES (10, 1), (20, 2), (30, 3), (40, 4), (50, 99)")
+	mcmp.Exec(`select t.a from (select t1.id2, t2.id3, (select id2 from t1 order by id2 limit 1) as a from t1 join t2 on t1.id1 = t2.id4) t`)
+	mcmp.Exec(`SELECT COUNT(*) FROM (SELECT DISTINCT t1.id1 FROM t1 JOIN t2 ON t1.id1 = t2.id4) dt`)
```
Comment (systay) [inline go/test/endtoend/vtgate/queries/subquery/subquery_test.go:189]: This last test will fail without the changes in this PR:

```
Error: Received unexpected error:
target: ks_union.-80.primary: vttablet: rpc error: code = NotFound desc = Unknown column 't2.id4' in 'field list' (errno 1054) (sqlstate 42S22) (CallerID: userData1): Sql: "select distinct t2.id4, 1, dt.`t2.id4` from (select t2.id4 as `t2.id4` from t2) as dt", BindVars: {}
target: ks_union.80-.primary: vttablet: rpc error: code = NotFound desc = Unknown column 't2.id4' in 'field list' (errno 1054) (sqlstate 42S22) (CallerID: userData1): Sql: "select distinct t2.id4, 1, dt.`t2.id4` from (select t2.id4 as `t2.id4` from t2) as dt", BindVars: {} (errno 1054) (sqlstate 42S22) during query: SELECT COUNT(*) FROM (SELECT DISTINCT t1.id1 FROM t1 JOIN t2 ON t1.id1 = t2.id4) dt
```

**Review by harshit-gangal — APPROVED** [review]

**Pull request merged**